proposals/0125-incremental-accounts-hash.md

```diff
@@ -0,0 +1,391 @@
+4. Snapshot.hash of all accounts: incremental
+
+### Perfomance
+
+Some performance result - rust code from Solana runtime.
+
+|  | 10_000 accounts | 100_000 accounts |
+| --- | --- | --- |
+| #[bench] test_accounts_delta_hash (unmodified) | 2.2ms | 25ms |
+| without accumulate_account_hashes | 1.559ms | 17ms |
+| with blake3_xof (2048 bytes) | 1.563ms | 18ms |
+
+The first row shows the existing bench run “as is”, with 100k accounts (as in
+the current bench) or 10k (trivial change to the bench code).
+
+The second row shows that 70% of the computation is blake3, and therefore 30% is
+sort + Merkle tree.
+
+The third row shows the cost of switching from blake3 (256-bit) to blake3_xof
+(2048-bytes) that’s needed for the incremental hash. The difference is minimal.
+Accumulating the hashes is also negligible (it’s only 1024 u16 adds, and they
+get vectorized).
+
+**What does this mean for the incremental hash and for a transition period with
+both delta and incremental hash?**
+
+- The cost of the incremental hash would be roughly 1.4-1.5x, i.e. 3ms at 10k
+  accounts
+    - We need to compute a hash of the account before (0.7x) and after (0.7x)
+      executing the tx
+    - As explained above, this assumes we won’t store/cache any hash at all
+    - Some of this cost can be easily hidden, for example by computing the
+      account hash while  loading the account, in parallel to executing the tx
+      (this means that the total latency could be reduced to 70%)
+- During a transition period where both hashes are computed:
+    - Trivial: the cost is 2.5x
+    - Reusing the long blake3 computation for both hashes can reduce to 1.7x
+      (but probably not worth if if we’re dealing with 10k accounts and the
+      transition period is not super long)
+
+### Implementation Notes
+
+#### Efficient Caching of BLAKE3 State for Extending Output (Optional)
+
+Implementors of this specification may wish to cache or store the hash of the
+accounts. Doing this for the full 2KiB of BLAKE3 extended output could be
+prohibitively expensive. 
+
+To avoid this high storage cost we recommend retaining the BLAKE3 state
+necessary to recompute the extended output. This is comprised of:
+
+- Root input chaining value (32 bytes): The chaining value used as input to the
+  root compression function in the BLAKE3 algorithm.
+- The input data length in bytes modulo 64 bytes (1 byte): Note, if the
+  `data_len % 64 == 0 && data_len != 0` then this length is 64. Alternatively,
+  implementors already have the total input data length, and can just retain
+  this, computing the modulo on the fly.
+- Given the length value above, retain the last data_len modulo 64 bytes (note
+  the exception above) of the input (max 64-bytes).
+
+Together this state can be used to recompute the 2KiB extended BLAKE3 hash. [4,
+Section 2.6]
+
+## Impact
+
+This change only impacts validators. It'll be a transparent change for any other
+user.
```

Comment:
Yep, removing the accounts delta root makes it impossible to watch for changes in an account's state without a trusted 3rd party. This breaks the existing light client design that Tinydancer describes [here](https://x.com/tinydancerio/status/1791519024721072325), and would also prevent rollups from having secure bridges with the base layer. Such a bridge is being actively developed by Zeta Markets, so this would force them to scrap their plans for an L2 on Solana.

Reply:
FYI, this is currently on hold, unless surprises we'll resume work after Breakpoint and re-evaluate.

As for TinyDancer and (I learn now) Zeta Markets, I'd pay extra attention to edge cases, for example accounts with 0 lamports that disappear from bank hash (in the current Solana implementation, independently from this proposal).
Unlike Ethereum that was designed with ZK & light clients in mind, Solana was not. So I'd be really surprise to see that "everything works fine", even in obscure edge cases.

When we originally discussed this work we agreed that scalability was a priority wrt light clients, and that if we want light clients in Solana we should have a proper spec, not just assume that what exists is working. But as I said, this work is currently paused and when we'll resume it we'll definitely re-evaluate the situation (and, of course, will alert in this PR). Thanks for the new info!

Reply:
Any update on this now that Breakpoint is around the corner?